src/smart.ts

```diff
@@ -134,6 +136,60 @@ function any(tasks: Task[]): Promise<any> {
     });
 }
 
+/**
+ * The maximum length for a code verifier for the best security we can offer.
+ * Please note the NOTE section of RFC 7636 § 4.1 - the length must be >= 43,
+ * but <= 128, **after** base64 url encoding. This means 32 code verifier bytes
+ * encoded will be 43 bytes, or 96 bytes encoded will be 128 bytes. So 96 bytes
+ * is the highest valid value that can be used.
+ */
+ var RECOMMENDED_CODE_VERIFIER_LENGTH = 96;
+
+ /**
+  * Character set to generate code verifier defined in rfc7636.
+  */
+ var PKCE_CHARSET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
+
+ /**
+  * Generates a code_verifier and code_challenge, as specified in rfc7636.
+  */
+ function generatePKCECodes(env: fhirclient.Adapter) {
+   var output = new Uint32Array(RECOMMENDED_CODE_VERIFIER_LENGTH);
+   crypto.getRandomValues(output);
+   var codeVerifier = base64urlEncode(env, Array.from(output).map(function (num) {
+     return PKCE_CHARSET[num % PKCE_CHARSET.length];
```

Comment:
Again assuming we need a dependency like https://github.com/cisco/node-jose anyway, we can use `jose.util.randomBytes(RECOMMENDED_CODE_VERIFIER_LENGTH)` and pass the digest into `jose.util.base64url.encode`.